pkg/ceph/osd/health.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2017 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package osd
+
+import (
+	"time"
+
+	"github.com/rook/rook/pkg/ceph/client"
+	"github.com/rook/rook/pkg/clusterd"
+)
+
+var (
+	healthCheckInterval = 60 * time.Second
+
+	// waitThreshold is period before an action is taken
+	waitThreshold = 300 * time.Second
+)
+
+// Monitor defines OSD process monitoring
+type Monitor struct {
+	context *clusterd.Context
+	agent   *OsdAgent
+
+	// lastStatus keeps track of OSDs status
+	// key - Process ID; value: time of the status change.
+	lastStatus map[int]time.Time
+}
+
+// NewMonitor instantiates OSD monitoring
+func NewMonitor(context *clusterd.Context, agent *OsdAgent) *Monitor {
+	return &Monitor{context, agent, make(map[int]time.Time)}
+}
+
+// Run runs monitoring logic for osd processe's status at set intervals
+func (m *Monitor) Run(stopCh chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			logger.Infof("Stopping OSD monitoring.")
+			return
+
+		case <-time.After(healthCheckInterval):
+			logger.Debug("Checking osd processes status.")
+			err := m.procStatus()
+			if err != nil {
+				logger.Infof("Failed OSD status check: %s", err)
+			}
+		}
+	}
+}
+
+func (m *Monitor) procStatus() error {
+	osdDump, err := client.GetOSDDump(m.context, m.agent.cluster.Name)
+	if err != nil {
+		return err
+	}
+
+	// TODO: handling of OSDs that are not tracked by Agent or OSDDump
+	logger.Debugf("OSDDump vs. Agent.osdProc : %d", len(osdDump.OSDs)-len(m.agent.osdProc))
```

Comment:
i don't think this debug statement is going to be very useful.  `osdDump` is going to be a dump of all OSDs and `agent.osdProc` is going to be only the OSDs that this specific pod is running.  there should always be a difference when there is more than 1 OSD pod in the cluster.

Also, just knowing the numerical difference isn't that useful.  If you really wanted to compare them, and since it's DEBUG level, i would print out the whole structs, using `%+v`.

Reply:
Just had it there temporarily as I was observing differences during testing.
Will remove, using OSDs tracked by the agents as an authoritative source sounds good.